84-sarcastic-fringehead/technical_specification.md

```diff
@@ -0,0 +1,478 @@
+   - FIS returns the list of `FileUploads` corresponding to the `FileUploadReports`
+3. (`GET`) Serve `InspectionResults` for a given file ID
+   - A token-authenticated request supplies a file ID as a path parameter.
+   - FIS finds the existing `FileUploadReport` in its database, raising an error if it doesn't find it (should be translated to a 404 response but logged within the service as an error).
+   - Returns the subset of the `FileUploadReport` corresponding to the `InspectionResults` model structure. This is done to reduce the number of places that the secret ID is communicated, as well as the irrelevant box ID.
+4. (`POST`) Accept interrogation results
+   - A token-authenticated request contains information about a completed file interrogation. The information includes the following fields:
+     - `file_id`
+     - `inspection_result`
+     - If successful, then also:
+       - `storage_alias` of the interrogation bucket
+       - `encrypted_part_size`
+       - `encrypted_parts_md5`
+       - `encrypted_parts_sha256`
+   - FIS finds the existing `FileUploadReport` in its database (or raises an error).
+   - FIS updates the `FileUploadReport` with the received information and publishes this as a *persistent event*. 
+
+For a given file upload, the endpoints will be called in the order listed above.
+
+#### FIS Configuration
+The FIS needs the following configuration:
+- MongoDbConfig
+- KafkaConfig
+- ApiConfigBase
+- LoggingConfig
+- OpenTelemetryConfig
+- EventPubConfig:
+  - ghga-event-schemas -> FileUploadReportEventsConfig
+- OutboxSubConfig:
+  - ghga-event-schemas -> FileUploadEventsConfig
+- ekss_api_url
+- wkvs_api_url
+
+Upon startup, the HTTP API instance of the FIS will retrieve the list of Data Hub public keys from the WKVS.
+
+In addition, the existing functionality and config that directly interacts with Vault should be removed so that EKSS is the sole middleman for Vault activity.
+
+
+#### Open questions
+1. What do we do with the existing FIS data? Just leave it there for now?
+   - Persisted events stored in the FIS are analogous to FileUploadReport events, we might be able to migrate them for continuity.
+
+#### Work to be performed for the FIS
+- [ ] Ensure DLQ is enabled
+- [ ] Add outbox subscriber for `FileUpload` events
+- [ ] Add persistent publisher for `FileUploadReport` events
+- [ ] Add HTTP endpoints as outlined above
+- [ ] Write migration for existing persisted events if possible
+- [ ] Write tests
+
+
+### DHFS:
+The DHFS is a new service that is operated by the Data Hubs for the purpose of performing file validation and re-encryption, and to keep file ingest in general as a federated operation. The DHFS operates two instances: a client instance, which performs the interrogation work and is always running; and a cleanup instance, which runs at some interval and deletes files from the `interrogation` bucket once they've been copied to permanent storage.
+
```

Comment:
You should mention here that the Data Hubs do not share the event streams with central, so the DFHS cannot subscribe to FileUploads directly and needs to poll a FileUpload list from central via REST  instead.